lua/entities/gmod_wire_expression2/core/custom/constraintcore.lua

```diff
@@ -350,91 +353,103 @@ end
 
 
 --  == Hydraulic ==
-
--- Note: Winch is just a rename of Hydraulic with the last parameter True.
---- Makes a winch constraint (stored at index <index>) between <ent1> and <ent2>, at vectors local to their respective ents, with <width> width.
-e2function void winch(index, entity ent1, vector v1, entity ent2, vector v2, width)
+local function createHydraulic(self, index, ent1, ent2, v1, v2, width, bone1, bone2, constant, damping, rdamping, mat, stretch, color)
 	if not checkEnts( self, ent1, ent2 ) then return end
-	if not checkCount( self, "Hydraulic" , ent1, ent2 ) then return end
+	if not checkCount( self, "Hydraulic", ent1, ent2 ) then return end
 	if not checkEdicts( self ) then return end
 	local constraints = setupEntConstraints( ent1 )
 
-	if width < 0 or width > 50 then width = 1 end
+	width = math.Clamp(width, 0, 50)
 
 	local existing = constraints[index]
 	if IsValid( existing ) then existing:Remove() end
-
-	local constant, dampen = CalcElasticConsts( ent1:GetPhysicsObject(), ent2:GetPhysicsObject(), ent1, ent2 )
-	local cons, rope = constraint.Elastic( ent1, ent2, 0, 0, v1, v2, constant, dampen, 0, "cable/cable2", width, true )
+	
+	if color ~= nil then
+		color = Color(color[1],color[2],color[3],255)
+	end
+	if not constant or not damping then
+		if bone1 then
+			phys1 = getBone(ent1,bone1)
+		else
+			phys1 = ent1:GetPhysicsObject()
+		end
+		if bone2 then
+			phys2 = getBone(ent2,bone2)
+		else
+			phys2 = ent2:GetPhysicsObject()
+		end
+		
+		constant, damping = CalcElasticConsts( phys1, phys2, ent1, ent2 )
+	end
+	
+	local cons, rope = constraint.Elastic( ent1, ent2, bone1 or 0, bone2 or 0, v1, v2, constant, damping, rdamping or 0, mat ~= "" and mat or "cable/cable2", width or 1, stretch ~= 0, color )
 	if not verifyConstraint( self, cons ) then return end
 
 	constraints[index] = cons
 	postCreate( self, "Hydraulic", ent1, ent2, cons, rope )
 end
 
---- Makes a hydraulic constraint (stored at index <index>) between <ent1> and <ent2>, at vectors local to their respective ents, with <width> width.
-e2function void hydraulic(index, entity ent1, vector v1, entity ent2, vector v2, width)
-	if not checkEnts( self, ent1, ent2 ) then return end
-	if not checkCount( self, "Hydraulic", ent1, ent2 ) then return end
-	if not checkEdicts( self ) then return end
-	local constraints = setupEntConstraints( ent1 )
+e2function void hydraulic(index, entity ent, vector v1, bone bone, vector v2, width)
+	entBone, index = isValidBone(bone)
```

Comment:
Missing `local` for all of these.

Reply:
Hah and that's why we have computers generate functions for us.